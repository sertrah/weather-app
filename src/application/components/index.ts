export { default as AppLayout } from "./AppLayout";
export { default as Loader } from "./Loader";
export { default as NotificationMessage } from "./NotificationMessage";
export { default as Container } from "./Container";
export { default as LineChart } from "./LineChart";
export { default as BarChart } from "./BarChart";
export { default as RangeSlider } from "./RangeSlider";
