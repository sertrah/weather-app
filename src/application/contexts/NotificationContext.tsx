import React, {
  createContext,
  useState,
  useMemo,
  useCallback,
} from "react";
import NotificationMessage from "../components/NotificationMessage";

type ContextProps = any;

const NotificationContext = createContext<Partial<ContextProps>  | null>(null);

const defaultConfig = {
  duration: 3000,
  mountingElement: document.body,
};
const removeById = (id: number) => (prev: any) =>
  prev.length > 0 ? prev.filter((a: any) => a.id !== id) : prev;

const  NotificationProvider = ({ children, config }: any)  => {
  const [notificationList, setNotificationList] = useState<{id: number, title?: string, message?: string}[]>([]);
  const { duration = 3000, mountingElement = document.body } = {
    ...defaultConfig,
    ...config,
  };

  const handleClose = useCallback(
    (id) => {
      clearTimeout(id);
      setNotificationList(removeById(id));
    },
    [setNotificationList]
  );

  const notify = useMemo(() => {
    const notifyFn = (notification: any) => {
      const id: any = setTimeout(() => {
        setNotificationList(removeById(id));
      }, duration);
      setNotificationList((list) => [...list, { id, ...notification }]);
    };

    const notifyStatus = (status: any) => (notification: any) =>
      notifyFn({ ...notification, status });

    notifyFn.success = notifyStatus("success");
    notifyFn.error = notifyStatus("error");

    return notifyFn;
  }, [setNotificationList, duration]);

  return (
    <>
      <NotificationContext.Provider value={notify}>
        {children}
      </NotificationContext.Provider>
      <NotificationMessage
        notifications={notificationList}
        onClose={handleClose}
        mountingElement={mountingElement}
      />
    </>
  );
}

export { NotificationContext, NotificationProvider };
