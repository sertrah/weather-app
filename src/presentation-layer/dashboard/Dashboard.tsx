import React, { FunctionComponent, useState, useMemo } from "react";
import { useQuery } from "react-query";

import ChanceRainChart from "./ChanceRainChart";
import AmountRainFallChart from "./AmountRainFallChart";
import { RangeSlider, Container, Loader } from "application/components";
import { weatherService } from "infraestructure/services";
import { useNotification } from "application/hooks";

function chanceOfRain(pressure: number, temperature: number, amount: number) {
  var score =
    Math.log(amount + 1) * Math.log(pressure - 929) * Math.log(temperature - 9);
  var mean = Math.min(Math.max(score, 0), 100);
  var upper_bound = Math.min(1.5 * mean, 100);
  var lower_bound = Math.max(0.5 * mean, 0);
  return [lower_bound, mean, upper_bound];
}

const Dashboard: FunctionComponent = () => {
  const [pressure, setPressure] = useState(1030);
  const [temperature, setTemperature] = useState(30);
  const notify = useNotification();
 

  const { data: rainFallList, isLoading } = useQuery(
    "rainFallList",
    weatherService.getAll,
    {
      retry: 1,
      retryDelay: 3000,
      onError: (error)=> {
        notify.success({
          title: "Weather Service",
          message: error,
        });
      }
    }
  );
  const chanceOfRainCalculated = useMemo(() => {
    return (
      rainFallList &&
      rainFallList[0].days.map(({ amount, day }) => ({
        label: day,
        value: chanceOfRain(pressure, temperature, amount),
      }))
    );
  }, [pressure, temperature, rainFallList]);
  return (
    <Container>
      <h1>Adidas Technical Test</h1>
      {isLoading && <Loader />}
      <section className="dashboard__container">
        <RangeSlider
          title="Pressure [hPa]"
          max={1030}
          min={970}
          value={pressure}
          onChange={setPressure}
        />

        {chanceOfRainCalculated && (
          <ChanceRainChart weatherList={chanceOfRainCalculated} />
        )}
        <RangeSlider
          title="Temperature [Â°C]"
          max={35}
          min={10}
          value={temperature}
          onChange={setTemperature}
        />
        {rainFallList && (
          <AmountRainFallChart weatherList={rainFallList[0].days} />
        )}
      </section>
    </Container>
  );
};

export default Dashboard;
