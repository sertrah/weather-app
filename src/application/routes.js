import Dashboard from "presentation-layer/dashboard";

import { ROUTER_PATH_LIST } from "./constants";

const routes = [
  {
    path: ROUTER_PATH_LIST.default,
    component: Dashboard,
  },
];

export default routes;
