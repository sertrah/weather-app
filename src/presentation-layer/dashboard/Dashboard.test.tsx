import React from "react";
import { render, screen } from "@testing-library/react";
import { QueryClientProvider, QueryClient } from "react-query";
import { NotificationProvider } from "application/contexts";
import Dashboard from "./Dashboard";

const queryClient = new QueryClient();

describe("Dashboard tests", () => {
  it("renders Adidas title", () => {
    const component = render(
      <QueryClientProvider client={queryClient}>
        <NotificationProvider>
          <Dashboard />{" "}
        </NotificationProvider>
      </QueryClientProvider>
    )
    const textElement = screen.getByText(/Adidas Technical Test/i);
    expect(textElement).toBeInTheDocument();
  });

  it("renders well", () => {
    const component = render(
      <QueryClientProvider client={queryClient}>
        <NotificationProvider>
          <Dashboard />{" "}
        </NotificationProvider>
      </QueryClientProvider>
    );
    expect(component.container).toMatchSnapshot();
  });
});
