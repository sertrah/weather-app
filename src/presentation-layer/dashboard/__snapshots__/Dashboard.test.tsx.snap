// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard tests renders well 1`] = `
<div>
  <div
    class="container"
  >
    <h1>
      Adidas Technical Test
    </h1>
    <div
      class="loader"
    >
      <div
        class="loader--container"
      >
        <div
          class="loader--item"
        />
        <p />
      </div>
    </div>
    <section
      class="dashboard__container"
    >
      <div
        class="rangeSlider"
        style="width: 200px;"
      >
        <p
          class="rangeSlider--text"
        >
          Pressure [hPa]
        </p>
        <p
          class="rangeSlider--label"
        >
          1030
        </p>
        <input
          class="rangeSlider--input"
          max="1030"
          min="970"
          type="range"
          value="1030"
        />
      </div>
      <div
        class="rangeSlider"
        style="width: 200px;"
      >
        <p
          class="rangeSlider--text"
        >
          Temperature [Â°C]
        </p>
        <p
          class="rangeSlider--label"
        >
          30
        </p>
        <input
          class="rangeSlider--input"
          max="35"
          min="10"
          type="range"
          value="30"
        />
      </div>
    </section>
  </div>
   
</div>
`;
